[GSoD]  Added tutorials directory on gRPC-Gateway (#1829)
grpc-ecosystem/grpc-gateway: docs/docs/tutorials/simple_hello_world.md

---
layout: default
title: Creating a simple hello world with gRPC-Gateway
parent: Tutorials
nav_order: 4
---

### Creating a simple hello world with gRPC-Gateway

### gRPC-Gateway

We all know that gRPC is not a tool for everything. There are cases where we still want to provide a traditional RESTful JSON API. The reasons can range from maintaining backwards-compatibility to supporting programming languages or clients not well supported by gRPC. But coding another API for REST is quite a time consuming and tedious.

So is there any way to code just once, but can provide APIs in both gRPC and REST at the same time?

The answer is Yes.

### Usage

One way to achieve that is to use gRPC gateway. gRPC gateway is a plugin of the protocol buffer compiler. It reads the protobuf service definitions and generates a proxy server, which translates a RESTful HTTP call into gRPC request.

All we need to do is a small amount of configuration
in the service. Before start coding, we have to install some
tools.

> Use `go get -u` to download the following packages:

Reviewer: ```suggestion
We will be using a Go gRPC server in the examples, so please install Go first from [https://golang.org/dl/](https://golang.org/dl/).

After installing Go, use `go get` to download the following packages:
```

```sh

> go get -u github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway
> go get -u github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger
> go get -u github.com/golang/protobuf/protoc-gen-go

Reviewer: These are the old binaries, these are the updated links

```suggestion
go get github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-grpc-gateway
go get github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-openapiv2
go get google.golang.org/protobuf/cmd/protoc-gen-go
```

```

> 

Reviewer: ```suggestion

This installs the protoc generator plugins we need to generate the stubs. Make sure to add `$GOPATH/bin` to your `$PATH` so that executables installed via `go get` are available on your `$PATH`.
```

Reviewer: Everything above this line should on a separate page about "pre-requirements".

### Define and Generate proto files

Before we create a gRPC service, we should create a proto file to define what we need, here we create a file named `hello.proto` to show.

```proto
syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";

> option java_outer_classname = "HelloWorldProto";

Reviewer: We can remove these

package helloworld;

import "google/api/annotations.proto";

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {
        option (google.api.http) = {
        post: "/v1/example/echo"
        body: "*"
    };
  }
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

> ```

Reviewer: I wonder if we should use this hello.proto instead of the arithmetic's server example. It's simpler and easier to implement as an example.

Then we should generate gRPC stub via protoc

```sh
protoc -I/usr/local/include -I. \
-I$GOPATH/src \
-I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
--go_out=Mgoogle/api/annotations.proto=github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis/google/api,plugins=grpc:. \

> helloworld/helloworld.proto

Reviewer: Where did you get this from? Our docs haven't looked like this for a good while. Maybe we should just refer to `Generating the stubs` here instead of repeating ourselves?

```

The `helloworld.pb.go` file will be generated. So `helloworld.pb.go` is required by the server service.

Next we need to use protoc to generate the go files needed by the gateway.

```sh
protoc -I/usr/local/include -I. \
-I$GOPATH/src  -I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
--swagger_out=logtostderr=true:. \
helloworld/helloworld.proto

> ```

Reviewer: Lets move this to two separate files, one for generating the grpc-gateway with `protoc` and one for doing it with `buf`.

The `helloworld.pb.gw.go` file will be generated. This file is the protocol file used by gateway for protocol conversion between grpc and http.

After the protocol file is processed, the gateway code is needed.

### Implementation

The gateway code is as follows:

```go
package main

import (
    "flag"
    "net/http"

    "github.com/golang/glog"
    "github.com/grpc-ecosystem/grpc-gateway/runtime"
    "golang.org/x/net/context"
    "google.golang.org/grpc"

    gw "grpc-helloworld-gateway/helloworld"
)

var (
    echoEndpoint = flag.String("echo_endpoint", "localhost:50051", "endpoint of YourService")
)

func run() error {
    ctx := context.Background()
    ctx, cancel := context.WithCancel(ctx)
    defer cancel()

    mux := runtime.NewServeMux()
    opts := []grpc.DialOption{grpc.WithInsecure()}
    err := gw.RegisterGreeterHandlerFromEndpoint(ctx, mux, *echoEndpoint, opts)
    if err != nil {
        return err
    }

    return http.ListenAndServe(":8080", mux)
}

func main() {
    flag.Parse()
    defer glog.Flush()

    if err := run(); err != nil {
        glog.Fatal(err)
    }
}

> ```

Reviewer: Remove this. We will just reference the boilerplate repo instead.

First echoEndpoint stores the server information that needs to be connected, and then registers and binds the new server with RegisterGreeter Handler FromEndpoint in `gw.go`. Then the lower level will connect to the remote server address provided by echoEndpoint, so the gateway establishes the connection as the client and remote server, and then starts the new server with http, gateway. It serves as a server to provide HTTP services to the outside world.

This is the end of the code. Let's test it.

Start the greeter_server service first, and then start the gateway. Then gateway connects to greeter_server and establishes http monitoring.

Then we use curl to send http requests:

```sh
curl -X POST -k http://localhost:8080/v1/example/echo -d '{"name": " world"}
```

```
{"message":"Hello  world"}
```

The process is as follows:

curl sends a request to the gateway with the post, gateway as proxy forwards the request to greeter_server through grpc, greeter_server returns the result through grpc, the gateway receives the result, and json returns to the front end.

> In this way, the transformation process from http json to internal grpc is completed through grpc-gateway.

Reviewer: This is great, definitely keep this, but move it to a new page that talks about adding the grpc-gateway annotations to an existing protobuf file.

The grcp-gateway the server is created according to the `google.api.http` annotations in your service definitions.

HttpRule defines the gRPC / REST mapping scheme. The mapping defines how different parts of a gRPC request message are mapped to the URL path, URL request parameters, and HTTP request body. It also controls how the gRPC response message is displayed in the HTTP response body. HttpRule is usually specified as a `google.api.http` annotation in the gRPC method.

Each mapping defines a URL path template and an HTTP method. A path template can refer to one or more fields in a gRPC request message if each field is a non-repeating field with a primitive type. The path template controls how the request message fields are mapped to the URL path.

```proto
import "google/api/annotations.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
message StatusResponse {
 google.protobuf.Timestamp current_time = 1;
}
service MyService {
 rpc Status(google.protobuf.Empty)
  returns (StatusResponse) {
   option (google.api.http) = {
    get: "/status"
   };
  }
 }
}
```

> 

Reviewer: This would make a great separate "Learn more" page I think.

You will need to provide the necessary third-party `protobuf` files to the `protoc` compiler. They have included in the `grpc-gateway` repository in the `[third_party/googleapis](https://github.com/grpc-ecosystem/grpc-gateway/tree/master/third_party/googleapis)` folder, and we recommend copying them to the project file structure.

> You will need to provide the necessary third-party `protobuf` files to the `protoc` compiler. They are included in the `[grpc-gateway` repository](https://github.com/grpc-ecosystem/grpc-gateway/) in the `[third_party/googleapis](https://github.com/grpc-ecosystem/grpc-gateway/tree/master/third_party/googleapis)` folder and we recommend copying them to the project file structure.

Reviewer: What happened here? These sentences are identical.

---
layout: default
title: Creating a simple hello world with gRPC
parent: Tutorials
nav_order: 2
---

## Creating a simple hello world with gRPC

To understand the gRPC-Gateway we are going to first make a hello world gRPC service.

### Defining your gRPC service using protocol buffers

Before we create a gRPC service, we should create a proto file to define what we need, here we create a file named `hello_world.proto` in the directory `proto/helloworld/hello_world.proto`.

```proto

> syntax = "proto3";
> 
> package helloworld;
> 

Reviewer: We are missing the service definition, we need it in this file (without the `google.http.api` annotations):
```suggestion
syntax = "proto3";

package helloworld;

// Here is the overall greeting service definition where we define all our endpoints
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
}

```

---
layout: default
title: Creating a simple hello world with gRPC
parent: Tutorials
nav_order: 2
---

## Creating a simple hello world with gRPC

To understand the gRPC-Gateway we are going to first make a hello world gRPC service.

### Defining your gRPC service using protocol buffers

Before we create a gRPC service, we should create a proto file to define what we need, here we create a file named `hello_world.proto` in the directory `proto/helloworld/hello_world.proto`.

```proto
syntax = "proto3";

package helloworld;

// The greeting service definition
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}
}

// The request message containing the user's name
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

> ```

Reviewer: Can we add a bit more information here to say what we're actually doing? People coming here may be new to gRPC. Describe what the HelloRequest and HelloReply represent, what the types represent, and what the service represents.

---
layout: default
title: Creating a simple hello world with gRPC
parent: Tutorials
nav_order: 2
---

# Creating a simple hello world with gRPC

To understand the gRPC-Gateway we are going to first make a hello world gRPC service.

## Defining your gRPC service using protocol buffers

Before we create a gRPC service, we should create a proto file to define what we need, here we create a file named `hello_world.proto` in the directory `proto/helloworld/hello_world.proto`.

> The gRPC service is defined using [protocol buffers][pb]. To learn more about how to define a service in a `.proto` file see [Basics tutorial](https://grpc.io/docs/languages/go/basics/). For now, all you need to know is that both the server and the client stub have a `SayHello()` RPC method that takes a `HelloRequest` parameter from the client and returns a `HelloReply` from the server, and that the method is defined like this:

Reviewer: ```suggestion
The gRPC service is defined using [Google Protocol Buffers][https://developers.google.com/protocol-buffers]. To learn more about how to define a service in a `.proto` file see their [Basics tutorial](https://grpc.io/docs/languages/go/basics/). For now, all you need to know is that both the server and the client stub have a `SayHello()` RPC method that takes a `HelloRequest` parameter from the client and returns a `HelloReply` from the server, and that the method is defined like this:
```

> The gRPC service is defined using [Google Protocol Buffers][https://developers.google.com/protocol-buffers]. To learn more about how to define a service in a `.proto` file see their [Basics tutorial](https://grpc.io/docs/languages/go/basics/). For now, all you need to know is that both the server and the client stub have a `SayHello()` RPC method that takes a `HelloRequest` parameter from the client and returns a `HelloReply` from the server, and that the method is defined like this:

Reviewer: Doh, this was probably my fault.
```suggestion
The gRPC service is defined using [Google Protocol Buffers](https://developers.google.com/protocol-buffers). To learn more about how to define a service in a `.proto` file see their [Basics tutorial](https://grpc.io/docs/languages/go/basics/). For now, all you need to know is that both the server and the client stub have a `SayHello()` RPC method that takes a `HelloRequest` parameter from the client and returns a `HelloReply` from the server, and that the method is defined like this:
```

---
layout: default
title: Creating a simple hello world with gRPC
parent: Tutorials
nav_order: 2
---

## Creating a simple hello world with gRPC

> To understand gRPC-Gateway we are going to make hello world gRPC service which uses gRPC-Gateway.

Reviewer: ```suggestion
To understand the gRPC-Gateway we are going to make a hello world gRPC service which uses the gRPC-Gateway.
```

### Creating a basic protos with HTTP annotations

The annotations define how gRPC services map to the JSON request and response. When using protocol buffers, each RPC must define the HTTP method and path using the `google.api.http` annotation.

So you will need to add `import "google/api/http.proto";` to the gRPC proto file.

Now, let's add the HTTP annotations to our proto file.

```proto
syntax = "proto3";

package helloworld;

import "google/api/annotations.proto";

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {
        option (google.api.http) = {
        post: "/v1/example/echo"
        body: "*"
    };
  }
}

> ```

Reviewer: Remove mentions of the grpc-gateway in this section. The protofile can stay, but without the annotations.

---
layout: default
title: Creating a simple hello world with gRPC-Gateway
parent: Tutorials

> nav_order: 5

Reviewer: This should be the second page, this is a great simple introduction

---

> ### Creating a simple hello world with gRPC-Gateway

Reviewer: ```suggestion
### Creating a simple hello world with gRPC
```

### gRPC-Gateway

We all know that gRPC is not a tool for everything. There are cases where we still want to provide a traditional RESTful JSON API. The reasons can range from maintaining backwards-compatibility to supporting programming languages or clients not well supported by gRPC. But coding another API for REST is quite a time consuming and tedious.

So is there any way to code just once, but can provide APIs in both gRPC and REST at the same time?

> The answer is Yes.

Reviewer: We should move this to the introduction on the first page.

### Define and Generate proto files

> Before we create a gRPC service, we should create a proto file to define what we need, here we create a file named `hello.proto` to show.

Reviewer: ```suggestion
Before we create a gRPC service, we should create a proto file to define what we need, here we create a file named `hello_world.proto` in the directory `proto/helloworld/hello_world.proto`.
```

```proto
syntax = "proto3";

package helloworld;

import "google/api/annotations.proto";

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {

>         option (google.api.http) = {
>         post: "/v1/example/echo"
>         body: "*"
>     };

Reviewer: Remove this for now, this page will just be protobuf and grpc. We will add the http annotations in another file.
```suggestion
```

  }
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}
```

Then we should generate gRPC stub via protoc

```sh
protoc -I . \
   --go_out ./gen/go/ --go_opt paths=source_relative \
   --go-grpc_out ./gen/go/ --go-grpc_opt paths=source_relative \
   your/service/v1/hello_world.proto
```

The `helloworld.pb.go` file will be generated. So `helloworld.pb.go` is required by the server service.

Next we need to use protoc to generate the go files needed by the gateway.

```sh
protoc -I/usr/local/include -I. \
-I$GOPATH/src  -I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
--swagger_out=logtostderr=true:. \
helloworld/helloworld.proto
```

The `helloworld.pb.gw.go` file will be generated. This file is the protocol file used by gateway for protocol conversion between grpc and http.

> After the protocol file is processed, the gateway code is needed.

Reviewer: Remove this, the next page (generate with buf or protoc) will deal with generation.

### Implementation

For `main.go` file of hello world we can refer to the boilerplate repository [grpc-gateway-boilerplate

> Template](https://github.com/johanbrandhorst/grpc-gateway-boilerplate).

Reviewer: I think we will need to move this to a separate file that deals with creating `main.go` after we generated the files. This should be file number 4.
